pkgs/top-level/all-packages.nix

```diff
@@ -26878,8 +26878,7 @@ with pkgs;
 
   ocf-resource-agents = callPackage ../os-specific/linux/ocf-resource-agents { };
 
-  open-vm-tools = callPackage ../applications/virtualization/open-vm-tools { };
-  open-vm-tools-headless = open-vm-tools.override { withX = false; };
```

Comment:
The readme made me think that I needed to use `callPackage`. Is that incorrect?

Reply:
Let me explain.

Most of time a package in all-packages is declared as you have seen above:

```nix
package = callPackage path/to/package/default.nix { };
```

Moving it to by-name is straightforward: just move the file and delete the reference in all-packages.
A special procedure will "reinsert" the code below:

```nix
package = callPackage ../by-name/pa/package/package.nix { };
```

Sometimes there are  packages with multiple options:

```nix
{ lib, 
  stdenv,
  engine ? "SDL", # possible values: SDL X11 CURSES 
}:
```

Per convenience, we create aliases at all-packages.nix:

```nix
package = callPackage path/to/package/default.nix { };
packageSDL = package.overrides { engine = "SDL"; };
packageX11 = package.overrides { engine = "X11"; };
```

In the case above, you need to deal only with the migration of the first expression, since the two others below are mere references. Conceptually you would end with something like

```nix
package = callPackage ../by-name/pa/package/package.nix { };
packageSDL = package.overrides { engine = "SDL"; };
packageX11 = package.overrides { engine = "X11"; };
```

But the first line can be deleted since it will be added implicitly:

```nix
packageSDL = package.overrides { engine = "SDL"; };
packageX11 = package.overrides { engine = "X11"; };
```

Reply:
The case pointed out in the documentation is for a different situation:

```nix
package = callPackage path/to/package/default.nix { ffmpeg = ffmpeg_5; };
```

Usually an expression like that happens because the upstream developers of `package` did not update their code to work with the most recent `ffmpeg`.

In this case, the only method to migrate to by name is by keeping a call to `callPackage`.

```nix
package = callPackage ../by-name/pa/package/package.nix { ffmpeg = ffmpeg_5; };
```

because, obviously,

```nix
package = package.override { ffmpeg = ffmpeg_5; };
```

would cause a recursive loop.